ci/closures.sh

```
@@ -0,0 +1,18 @@
+#!/bin/bash
+set -x -o pipefail
+
+# Enable printing of all potential GC allocation sources to stdout
+export DFLAGS=-vgc
+
+# Ensure there are no lines about closure allocations in the output.
+# Note explicit check for `grep` exit status 1, i.e. no lines found.
+# This means tests will fail if any of these conditions are met:
+#
+#   * `make` fails (exit status 2, propagated by `set -o pipefail`)
+#   * `grep` fails (exit status 2)
+#   * `grep` finds `closure` (exit status 0)
+#
+# Note that `make` can in principle exit with status 1 too, but only
+# if the `-q` option is used, which should never be the case for our
+# builds, so we can assume exit status 1 only comes from `grep`.
```

I don't see how the exit status of the first part of the pipeline could propagate to `test`.

---

`set -o pipeline` is used.  So IIUC that means that if `(grep -e "closure"; test $? -eq 1)` has exit status 0, and `make` has a non-zero exit status, the latter will be the final exit status of the line.

---

It's just a cautionary note.  In one draft I was considering putting the `test` on a follow-up line, and checking the exit status of the entire pipeline; this should in principle be valid.

---

Yes. But in the comment you seem to argue that the exit code of make can somehow end up being used in the `test $? -eq 1`, don't you? My point is that the exit status of the make command cannot be used in the latter part of the pipeline. At least I don't see how it could happen for the pipeline as written. That's why I find the comment confusing. To me there are two parts of the pipeline. The latter part exits with `0` or `1` depending on the exit code returned by `grep`. The former part exits with whatever `make` exits and `pipefail` instructs the shell to return the exit code of rightmost failing part instead of just returning the exit code of the last command.